File: data_structures/c++/queue.cpp
#include <bits/stdc++.h>
using namespace std;

struct node
{
        int data;
        node *next;
};
struct node *head = NULL;

int display_peek()
{
        return head->data;
}

void display()
{
        struct node *temp;
        if (head == NULL)
        {
                printf("[*] Queue is empty");
                printf("-----------------------------------\n");
        }
        else
        {
                temp = head;
                printf("[+] Queue: ");
                while (temp != NULL)
                {
                        printf("%d ", temp->data);
                        temp = temp->next;
                }
                printf("\n-----------------------------------\n");
        }
}

void enqueue(int x){
        struct node *new_node = (struct node *)malloc(sizeof(struct node));
        if (head == NULL){
                new_node->data = x;
                new_node->next = NULL;
                head = new_node;
                printf("[^] queue started\n");
                printf("[+] peek: %d\n", display_peek());
                display();
        }
        else{
                printf("[^] %d added\n", x);
                printf("[+] still peek: %d\n", display_peek());
                new_node->data = x;
                new_node->next = NULL;
                struct node *temp = head;
                while (temp->next != NULL)
                {
                        temp = temp->next;
                }
                temp->next = new_node;
                display();
        }
}

void dequeue(){
        struct node *new_node = (struct node *)malloc(sizeof(struct node));
        if (head == NULL)
        {
                printf("[*] Queue is empty");
        }
        else
        {
                printf("[+] %d dequeued\n", display_peek());
                struct node *temp = head;
                head = head->next;
                temp->next = NULL;
                printf("[+] new peek: %d\n", display_peek());
                display();
        }
}

int main(){
        enqueue(1);
        enqueue(12);
        enqueue(3);
        enqueue(6);
        enqueue(111);
        dequeue();
        dequeue();
        dequeue();
}

File: data_structures/python/queue.py
import sys
stdwrite = lambda line: sys.stdout.write(line)

class Queue:
    class __Node:
        def __init__(self):
            self.data = None
            self.next = None

        def set_data(self, data):
            self.data = data

        def get_data(self):
            return self.data

        def set_next(self, next):
            self.next = next

        def get_next(self):
            return self.next

    __head = None

    def display(self):
        if not Queue.__head:
            stdwrite("Queue is empty\n")
        else:
            stdwrite("Queue: peek")
            temp = Queue.__head
            while temp:
                stdwrite(" -> "+str(temp.get_data()))
                temp = temp.get_next()
            stdwrite("\n")

    def enqueue(self, data):
        if not Queue.__head:
            new = self.__Node()
            new.set_data(data)
            Queue.__head = new
        else:
            new = self.__Node()
            new.set_data(data)
            temp = Queue.__head
            while temp.get_next():
                temp = temp.get_next()
            temp.set_next(new)
        self.display()

    def dequeue(self):
        if not Queue.__head:
            stdwrite("Queue is already empty\n")
        else:
            Queue.__head = Queue.__head.get_next()
            self.display()


if __name__ == '__main__':
    q = Queue()
    q.enqueue(1)
    q.enqueue(12)
    q.enqueue(32)
    q.enqueue(43)
    q.enqueue(232)
    q.enqueue(43252)
    q.dequeue()
    q.dequeue()






